### ISSUE TITLE
Cloud Storage tests fail frequently on CircleCI

### ISSUE THREAD

**Issue by remi** [GoogleCloudPlatform/ruby-docs-samples#82]
See failed Travis storage build:
https://travis-ci.org/GoogleCloudPlatform/ruby-docs-samples/jobs/163856203

The bucket being used for tests here appears to also be configured to store access logs.  Determine the cause.

Look into bucket (and project) isolation for test suites.  Note that tests currently run in parallel on Travis.

**Comment by remi** [comment]
Additional failing build.  Spec needs fix.

https://travis-ci.org/GoogleCloudPlatform/ruby-docs-samples/jobs/163856203

**Comment by remi** [comment]
Thought: each test suite can use bucket `[Project ID]-[Name of test suite]-[timestamp]`, eg. `"TestProject-BigQuery-20160115123059.0462"`

With our current Travis configuration...
 - Multiple runs of all test suites may happen at the same time (multiple builds of ruby-docs-samples in parallel, eg. multiple pull requests)
 - Multiple runs of test suites may happen at the same time within one overall build (as part of our *current* Travis matrix setup)

Vote: delete buckets `after :all` to not leave any buckets behind after each test suite run.  Find or create in `before` each example block.

**Comment by frankyn** [comment]
Does this mean we will have overhead of creating a bucket for each test? It shouldn't be too bad.

**Comment by remi** [comment]
There are some specs that create a clean bucket context before each spec, eg. [buckets_spec.rb](https://github.com/GoogleCloudPlatform/ruby-docs-samples/blob/c312fc534f912e9f5dc487e076d521e990b4333c/storage/spec/buckets_spec.rb#L28-L31)

I think I'd like to update the specs to look like...

```ruby
before do
  @test_bucket = create_test_bucket "BigQuery"
end

it "test that needs a second bucket" do
  @second_bucket = create_test_bucket "BigQuery second bucket"
end
```

Where `#create_test_bucket` is a required helper and the helper registers an after hook:

```ruby
# */spec_helpers/storage_helper.rb
require "spec_helpers/project_helper.rb"

def test_bucket_names
  @test_bucket_names ||= []
end

def create_test_bucket bucket_suffix
  bucket_suffix = bucket_suffix.downcase.gsub /[^\w+]/, "_"
  bucket_name = "#{default_project_id}-#{bucket_suffic}-#{Time.now.to_f}"
  test_bucket_names << bucket_name
  Google::Cloud.new.storage.create_bucket bucket_name
end

def delete_test_buckets!
  test_bucket_names.each do |bucket_name|
    bucket = Google::Cloud.new.storage.bucket bucket_name
    bucket.delete if bucket
  end
end

RSpec.configure do |config|
  config.after do
    delete_test_buckets!
  end
end
```

Or something.

Some of the other docs samples repos generate buckets for their test runs using a timestamp which is good for isolation.  Pretty simple.

**Comment by frankyn** [comment]
Gotcha, I like your idea for using RSpec after hooks :+1: 
Is it possible for concurrent tests to create the same bucket name? It only goes up to seconds.

**Comment by remi** [comment]
Time.now stores nanoseconds (since Epoch)

 - [Time#to_f](https://ruby-doc.org/core-2.3.0/Time.html#method-i-to_f)
 - [Time#nsec](http://ruby-doc.org/core-2.3.0/Time.html#method-i-nsec)

**Comment by frankyn** [comment]
Got it, I should have double checked that. 

Thanks!

**Comment by remi** [comment]
I have not proven that this is the reason for Storage samples, so I am renaming to focus on fixing all flakiness in Storage sample tests.

**Comment by remi** [comment]
Upgrading all flaky test failures to **`p0`**

Example of Storage failures:
 - [Build 266](https://circleci.com/gh/GoogleCloudPlatform/ruby-docs-samples/266?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)
 - [Build 263](https://circleci.com/gh/GoogleCloudPlatform/ruby-docs-samples/263)

**Comment by frankyn** [comment]
Fixed in #211
